[docs] Redesign DebOps Policies and Guidelines

drybjed opened the pull request:
This patch proposes a new design for DebOps Policies and Guidelines,
inspired by Python Enchancement Proposals. Some of the existing DebOps
policies are rewritten as DEPs.

drybjed committed:
[docs] Redesign DebOps Policies and Guidelines

This patch proposes a new design for DebOps Policies and Guidelines,
inspired by Python Enchancement Proposals. Some of the existing DebOps
policies are rewritten as DPGs.

drybjed commented:
Now I'm thinking that the `DEP-0001` should be something more akin to [PEP-1](https://www.python.org/dev/peps/pep-0001/) but I'm not sure if copying the whole thing and replacing/modifying parts relevant to DebOps is a good course of action... Opinions? Propably defining a framework for DEPs in the first one would be better than a list of keywords there.

drybjed committed:
[docs] Move DPG-* files to parent directory

This should help with browsing the DPG documents in the filesystem.

drybjed committed:
[docs] Update DPG-0001 to describe DPGs

The DPG-0001 is largely based on Python's PEP-1 document.

drybjed committed:
[docs] Update DPG headers to be consistent

drybjed committed:
[docs] Rename DPGs to DebOps Enhancement Proposals

The new name seems to be better suited for what these documents actually
are, and their purpose should be easier to understand for new
contributors. The contents will be updated in later commits.

drybjed committed:
Update DEP contents to reflect the name change

drybjed committed:
[ocs] Add a note on RFC 2119 use in specifications

drybjed committed:
[docs] Add DEP 12 with a DEP template

drybjed committed:
[docs] Make DEP numbers more human-readable

drybjed committed:
[docs] Add vim modeline in DEP sources

drybjed committed:
[docs] Update DEP header definition

The PEPs were designed with old format in mind, that was based on e-mail
headers. DEPs don't need to keep backwards compatibility with the old
formats and can start fresh with current best practices.

ypid commented on docs/dep/dep-0001.rst:
```diff
@@ -0,0 +1,599 @@
+  Track, Informational, or Process, and give it the status "Draft"
+
+Once the review process is complete, and the DEP editors approve it (note that
+this is *not* the same as accepting your DEP!), they will squash commit your
```
Why "squash commit", we don’t otherwise do this to my knowledge? I think we discussed this before, "squash committing" breaks signatures.

drybjed replied:
Good point. In Python as I understand it, this was done to keep the `git` repository with the PEPs clean. Original author of the PEP could have done multiple commits against PEP 9999, and after passing it off to the PEP editors they were free to renumber the PEP to the correct new number and squash all of the author commits together.

But we are tracking the commit origins via GPG signatures which breaks this system... I suppose that some of the current PEP workflow would need to be adapted for DEPs to handle this. I'll add a quick explanation on how to squash commits before submitting the work so that DEP authors could still have their work sighed with GPG keys, and there would need to be a mechanism to request new DEP numbers. I guess a reservation list in DEP 0 would suffice and DEP authors could be given the numbers after discussing an idea over the mailing list or GitHub issues.

ypid commented on docs/dep/dep-0001.rst:
```diff
@@ -0,0 +1,599 @@
+enhancement must be "debopsy" in order to be accepted by the DebOps
+maintainers. (However, "debopsy" is an imprecise term; it may be defined as
+whatever is acceptable to the DebOps maintainers. This logic is intentionally
+circular.).
```
debopsy :smile: Throwing in a joke here and there is a good idea with such long and otherwise welcoming precise specifications.

drybjed replied:
I cannot take the credit here, it's in the original PEP 1 as well. :-)

ypid commented on docs/dep/dep-0001.rst:
```diff
@@ -0,0 +1,599 @@
+   specification details can be emphasised by usage of the capitalized key
+   words, as described in the `RFC 2119`__.
+
+   .. __: https://tools.ietf.org/html/rfc2119
```
"must" (lower case) is used a lot in the document and the meaning is defined by RFC 2119. Although the RFC seems to be not 100 % precise about upper/lower case. I would say it is better to always write them in upper case (as done in the keyring docs and "older"/not newly written DEPs) so that it is clear that the author was aware of the meaning. That is also what the RFC implies and what it uses in itself.

Also because you write "details can be emphasised", if we don’t use upper case then, it is unclear.

drybjed replied:
[RFC 8174](https://tools.ietf.org/html/rfc8174) makes this more precise - only UPPERCASE words have the intended meaning defined by the RFC 2119.

In RFC 2119 there's a section about usage of the terms which states:

> Imperatives of the type defined in this memo must be used with care
> and sparingly.  In particular, they MUST only be used where it is
> actually required for interoperation or to limit behavior which has
> potential for causing harm

In my understanding it means that if you are writing a specification for an obvious thing that cannot be interpreted in multiple ways, using lowercase version is sufficient, as in speaking in normal voice. Only ambigious things require an uppercase version of the words which then carry **very specific meaning** defined by the RFC and that causes the given specification to be interpreted in a certain way, intended by the author.

ypid replied:
Works for me. To be explicit, we should the use:


```
      The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL
      NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
      "MAY", and "OPTIONAL" in this document are to be interpreted as
      described in BCP 14 [RFC2119] [RFC8174] when, and only when, they
      appear in all capitals, as shown here.
```

ypid committed:
[docs] Fix: Created header MUST be yyyy-mm-dd instead of SHOULD

Resolves specification conflict with: dep-0012.rst

ypid committed:
[docs] Fix spelling of DEPs

ypid commented on docs/dep/dep-0001.rst:
```diff
@@ -0,0 +1,599 @@
+changing their status).  See `DEP Editor Responsibilities & Workflow`_ for
+details. The current editors are:
+
+- Maciej Delmanowski
```
I might add myself when I have more time, but not right now, at least not officially. But I will try to keep up and should meet the DEP Editor requirements. But maybe don’t request a review from me where you think I am not involved in the discussion or which is unrelated to me, for now.

drybjed replied:
I'm aware of this and I did not intentionally add you or anybody else just yet, community is still small.

DEPs (and PEPs which they were based on) are designed for long-term discussions and mesh perfectly well with `git` history. This allows for interested parties to voice their concens at their own pace. But since this PR only introduces them, I wanted to have a few other people in the community review them beforehand to polish all the issues we might have.

ypid reviewed (changes requested):
Nice work! Pretty detailed and lets say, ready for the future :) Please find my, minor, fixes in: https://github.com/drybjed/debops/pull/6 and as comments.

ypid commented:
Defining DEPs is interesting and should help DebOps when it further grows. I approve the idea.

drybjed committed:
Merge branch 'ypid-#313-fixes' into docs-dpg

ypid reviewed (changes requested):
Thanks for merging. A few comments are open, otherwise I approve it.

drybjed committed:
[docs] Redesign DEP workflow around GPG signatures

DebOps uses GPG-signed commits in the 'git' repository which allow
better authentication of committers. DEP Workflow is redesigned to allow
new Draft DEPs to be merged into the repository with the commit
signatures preserved.

drybjed committed:
[docs] Clarify use of RFC 2119 key words in DEPs

ypid reviewed (approved):
LGTM.
